src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/JqlMemory.kt

```diff
@@ -3,5 +3,10 @@ package com.atlassian.performance.tools.jiraactions.api.memories
 import com.atlassian.performance.tools.jiraactions.api.page.IssuePage
 
 interface JqlMemory : Memory<String> {
```

Comment:
It seems like we want to treat JQLs not only as `String`s, but something more. We want to categorize them, tag them. We want a `Memory<Jql>`. This is why [primitive obsession](http://wiki.c2.com/?PrimitiveObsession) is a ticking time bomb. It strips us from extension points.

This is why the primitive obsession in `recallByTag` will also create problems in the future.

I propose to add a brand new interface, use domain classes (like `class Jql(val query: String, type: JqlType)`) and leave `JqlMemory` intact. The `AdaptiveJqlMemory` can implement both old and new ones at the same time.

Reply:
I considered this, but I'm not doing refactoring this time.

Reply:
@drspartez When making a tradeoff like this, it's a good practice to describe it in a commit's description. It adds a couple of extra benefits:
- You need to document the decision, so you need to make a logically sound block of text. For some persons (including me) it helps to sort things out. Sometimes it may affect a decision.
- Other people can review your decision. It'll help them to understand it and may help to find wholes in the reasoning. It may also affect the decision.
- If a change is confusing for the current readers, then it'll be confusing for future readers too. Commits are easier to access than PR.

Reply:
This is something new to me in scope of commit message. I was used to deal with project/design decisions on issue level. But here it is brainstormed more on PR. So i don't get why commit message appears here.

Reply:
> So i don't get why commit message appears here.

It appears here to indicate that we could avoid confusion like in this thread in the future if we document decisions. I also added reasons why it's a good practice to document it close to the source. 


decision documented in a commit description >> decision documented in the issue >> decision documented in the PR >> decision not documented at all.

For some cases, code comments would be best for documenting a decision. I'd leave it out of the scope of the discussion, as it highly depends on the context.

Reply:
@wyrzyk Would You like to propose neat pattern for such description relevant for commit message ?

Reply:
I'm not the best person in case of writing human-readable text, but I'll try:
```
JPERF-522: Add support for filtering by JQL type in JqlMemory

JqlMemory operates on Strings instead of JQL objects. It makes evolving the API
hard. I decided to keep using String for this change to avoid API refactoring costs.
```

Reply:
Cool. Thanks